services/httpd/handler_test.go

```diff
@@ -926,6 +927,93 @@ func TestHandler_XRequestId(t *testing.T) {
 	}
 }
 
+func TestThrottler_Handler(t *testing.T) {
+	t.Run("OK", func(t *testing.T) {
+		throttler := httpd.NewThrottler(2, 98)
+
+		// Send the total number of concurrent requests to the channel.
+		var concurrentN int32
+		concurrentCh := make(chan int)
+
+		h := throttler.Handler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			atomic.AddInt32(&concurrentN, 1)
+			concurrentCh <- int(atomic.LoadInt32(&concurrentN))
+			time.Sleep(1 * time.Millisecond)
+			atomic.AddInt32(&concurrentN, -1)
+		}))
+
+		// Execute requests concurrently.
+		const n = 100
+		for i := 0; i < n; i++ {
+			go func() { h.ServeHTTP(nil, nil) }()
+		}
+
+		// Read the number of concurrent requests for every execution.
+		for i := 0; i < n; i++ {
+			if v := <-concurrentCh; v > 2 {
+				t.Fatalf("concurrent requests exceed maximum: %d", v)
+			}
+		}
+	})
+
+	t.Run("ErrTimeout", func(t *testing.T) {
+		throttler := httpd.NewThrottler(2, 1)
+		throttler.EnqueueTimeout = 1 * time.Millisecond
+
+		resp := make(chan struct{})
+		h := throttler.Handler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			resp <- struct{}{}
+		}))
+
+		// First two requests should execute immediately.
+		go func() { h.ServeHTTP(nil, nil) }()
+		go func() { h.ServeHTTP(nil, nil) }()
+		time.Sleep(100 * time.Millisecond)
+
+		// Third request should be enqueued but timeout.
+		w := httptest.NewRecorder()
+		h.ServeHTTP(w, nil)
+		if w.Code != http.StatusServiceUnavailable {
+			t.Fatalf("unexpected status code: %d", w.Code)
+		} else if body := w.Body.String(); body != "request throttled, exceeds timeout\n" {
+			t.Fatalf("unexpected response body: %q", body)
+		}
+
+		// Allow 2 existing requests to complete.
+		<-resp
+		<-resp
+	})
+
+	t.Run("ErrFull", func(t *testing.T) {
+		throttler := httpd.NewThrottler(2, 1)
+
+		resp := make(chan struct{})
+		h := throttler.Handler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			resp <- struct{}{}
+		}))
+
+		// First two requests should execute immediately and third should be queued.
+		go func() { h.ServeHTTP(nil, nil) }()
+		go func() { h.ServeHTTP(nil, nil) }()
+		go func() { h.ServeHTTP(nil, nil) }()
+		time.Sleep(100 * time.Millisecond)
```

Comment:
This one is tricky. I have no solution besides perhaps some hooks in the throttler that can be set in tests to provide internal details, which also has downsides.

Reply:
I updated it so it using a much longer delay for CI builds. Exposing internals just for tests seems dirty to me. :) See 8a74c67.